### PR TITLE
Doc and plot

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

- [x] update documentation
- [x] clean up logging
- [x] reduce hash-map lookups
- [x] reduce number of memory allocations
- [x] generation of csv files for plotting
- [x] update Jenkinsfile to activate csv generation

The csv file generation, for now, can be toggled at compile time. The idea is that autopas should not be used at debug log level. Even when generating all csv files the data size is way smaller than the full log: For 500 iterations 500K vs 95K.

By default, csv files are not generated. We probably should enable them at least in an address sanitizer stage in Jenkins to test them. More detailed unit tests are not planned.

## ~Related Pull Requests~

- #PR

## ~Resolved Issues~

# How Has This Been Tested?

- [x] existing tests
- [x] Tested csv File generation in MPI context with ls1

### CODE REVIEW

**Review by SteffenSeckler — APPROVED** [review]
only minor things

File: cmake/modules/logging.cmake [hunk 1]
```diff
@@ -1,16 +1,49 @@
...
+option(AUTOPAS_COLORED_LOGGING "Print colored logging messages (only applies to cout and cerr)." OFF)
 if (AUTOPAS_COLORED_LOGGING)
     target_compile_definitions(autopas PUBLIC AUTOPAS_COLORED_CONSOLE_LOGGING)
     message(STATUS "Colored log messages enabled.")
 endif ()
+
+# option for GaussianClusterLogger
+option(AUTOPAS_Log_GaussianCluster "Generate a csv file about the gaussian cluster model that can be used for plotting." OFF)
+if (AUTOPAS_Log_GaussianCluster OR AUTOPAS_Log_All)
+    target_compile_definitions(autopas PUBLIC AUTOPAS_Log_GaussianCluster)
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply cmake/modules/logging.cmake:19]: no particular reason. Do you only want the definition / macro allcaps or also the cmake option? We are not really consistent on those anyways ¯\\\_(ツ)\_/¯

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply cmake/modules/logging.cmake:19]: as you prefer

File: Jenkinsfile [hunk 2]
```diff
@@ -202,7 +202,7 @@ pipeline {
                     steps {
                         container('autopas-clang6-cmake-ninja-make') {
                             dir("build-clang-ninja-addresssanitizer-debug") {
-                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Debug -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
+                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Debug -DDAUTOPAS_LOG_ALL=ON -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
```
Comment (SteffenSeckler) [inline Jenkinsfile:205]:
```suggestion
                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Debug -DAUTOPAS_LOG_ALL=ON -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
```

File: Jenkinsfile [hunk 3]
```diff
@@ -213,7 +213,7 @@ pipeline {
                     steps {
                         container('autopas-clang6-cmake-ninja-make') {
                             dir("build-clang-ninja-addresssanitizer-release") {
-                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Release -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
+                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Release -DDAUTOPAS_LOG_ALL=ON -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
```
Comment (SteffenSeckler) [inline Jenkinsfile:216]:
```suggestion
                                sh "CXXFLAGS=-Wno-pass-failed CC=clang CXX=clang++ cmake -G Ninja -DCCACHE=ON -DCMAKE_BUILD_TYPE=Release -DAUTOPAS_LOG_ALL=ON -DAUTOPAS_ENABLE_ADDRESS_SANITIZER=ON .."
```

File: src/autopas/AutoPas.h [hunk 4]
```diff
@@ -804,5 +811,11 @@ class AutoPas {
    */
   bool _externalMPICommunicator{false};
 
+  /**
+   * Suffix for all output files produced by this instance of AutoPas, e.g. from csv loggers.
+   * This is useful especially when multiple instances of AutoPas exist, especially in an MPI context.
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h:816]:
```suggestion
   * This is useful when multiple instances of AutoPas exist, especially in an MPI context.
```

File: src/autopas/AutoPas.h [hunk 5]
```diff
@@ -665,6 +665,13 @@ class AutoPas {
   MPI_Comm getMPICommunicator() { return _autopasMPICommunicator; }
 #endif
 
+  /**
+   * Suffix for all output files produced by this instance of AutoPas, e.g. from csv loggers.
+   * This is useful especially when multiple instances of AutoPas exist, especially in an MPI context.
```
Comment (SteffenSeckler) [inline src/autopas/AutoPas.h:670]:
```suggestion
   * This is useful when multiple instances of AutoPas exist, especially in an MPI context.
```

File: src/autopas/utils/logging/IterationLogger.cpp [hunk 6]
```diff
@@ -8,9 +8,9 @@
...
 #include "utils/Timer.h"
 
-autopas::IterationLogger::IterationLogger() {
-#ifdef AUTOPAS_Log_Iterations
-  auto outputFileName("AutoPas_iterationPerformance_" + utils::Timer::getDateStamp() + ".csv");
+autopas::IterationLogger::IterationLogger(const std::string &outputSuffix) {
+#ifdef AUTOPAS_LOG_ITERATIONS
+  auto outputFileName("AutoPas_iterationPerformance_" + outputSuffix + utils::Timer::getDateStamp() + ".csv");
   // Start of workaround: Because we want to use an asynchronous logger we can't quickly switch patterns for the header.
   // create and register a non-asychronous logger to write the header
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/IterationLogger.cpp:15]:
```suggestion
  // Create and register a non-asychronous logger to write the header.
```

File: src/autopas/utils/logging/IterationLogger.h [hunk 7]
```diff
@@ -52,8 +53,9 @@ class IterationLogger {
 
  private:
   static inline std::string loggerName() { return "IterationLogger"; };
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/IterationLogger.h:55]: suffix should go in here as well, in case we have two autopas instances on the same process

**Review by SteffenSeckler — CHANGES_REQUESTED** [review]
the suffix should also go into the logger names, as otherwise two autopas objects cannot exist on the same process.

**Comment by SteffenSeckler** [comment]
wow, messages from lgtm!

File: src/autopas/utils/logging/PredictionLogger.h [hunk 8]
```diff
@@ -45,7 +45,9 @@ class PredictionLogger {
                          size_t predictionErrorValue, size_t tuningPhaseCounter);
 
  private:
-  static inline std::string loggerName() { return "PredictionLogger"; };
+#ifdef AUTOPAS_LOG_PREDICTIONS
+  std::string _loggerName;
+#endif
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/PredictionLogger.h:50]: remove ifdef?

File: src/autopas/utils/logging/PredictionLogger.cpp [hunk 9]
```diff
@@ -8,11 +8,12 @@
 
 #include "autopas/utils/Timer.h"
 
-autopas::PredictionLogger::PredictionLogger(const std::string &outputSuffix) {
+autopas::PredictionLogger::PredictionLogger(const std::string &outputSuffix)
+    : _loggerName("PredictionLogger" + outputSuffix) {
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/PredictionLogger.cpp:12]: is not included in the ifdef => fails.

File: src/autopas/utils/logging/TuningDataLogger.h [hunk 10]
```diff
@@ -46,7 +46,9 @@ class TuningDataLogger {
                      size_t reducedValue, size_t smoothedVale);
 
  private:
-  static inline std::string loggerName() { return "TuningDataLogger"; };
+#ifdef AUTOPAS_LOG_TUNINGDATA
+  std::string _loggerName;
+#endif
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/TuningDataLogger.h:51]: remove ifdef

File: src/autopas/utils/logging/TuningResultLogger.h [hunk 11]
```diff
@@ -42,7 +42,9 @@ class TuningResultLogger {
   void logTuningResult(const autopas::Configuration &configuration, size_t iteration, long timeTuning);
 
  private:
-  static inline std::string loggerName() { return "TuningResultLogger"; };
+#ifdef AUTOPAS_LOG_TUNINGRESULTS
+  std::string _loggerName;
+#endif
```
Comment (SteffenSeckler) [inline src/autopas/utils/logging/TuningResultLogger.h:47]: remove ifdef

**Review by SteffenSeckler — APPROVED** [review]
good now, but doesn't compile, as _loggerName is in an ifdef, but accessed without an ifdef (in the constructor)

**Pull request merged**